setuptools/validation.py

```diff
@@ -0,0 +1,106 @@
+"""
+Validation logic for package metadata, supporting PEP 566.
+See: https://www.python.org/dev/peps/pep-0566/
+"""
+
+
+def single_line_str(val):
+    """A string that does not contain newline characters"""
+    if val is None:
+        return
+    assert isinstance(val, str), \
+        "is of type %s but should be of type str" % type(val)
+    assert '\n' not in val, \
+        "to contain a newline character (not allowed)"
+
+
+def multi_line_str(val):
+    """A string that can contain newline characters"""
+    if val is None:
+        return
+    assert isinstance(val, str), \
```

Comment:
Do not use `assert` for control flow. It is disabled when python is run optimized.

Reply:
What would be similarly elegant? I don't want to do an `if foo: raise Bar()` in all that many places.

Reply:
`assert` is not intended to be an "elegant" shorthand for raising an exception, it's a way of programmatically making assertions about how your code should work. You use it for things like enforcing a function's contract or documenting assumptions made by the programmer (which are then validated at runtime).

In this specific case, you cannot use it because it *will not work*. Assertions are to be treated as *optional* statements and if you run Python in optimized mode, they will simply not run at all, for example:

```bash
$ python -Oc "assert False"
$ python -c "assert False"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError
```

In this particular case, I would either raise `ValueError` or convert validators to functions returning a boolean, depending on the details. I'm not sure it's worth working out the details in this case, though, because as mentioned elsewhere this functionality rightly belongs in another library.

Reply:
@bittner Please update the code and use raise exception. Assert should be used only inside tests and I am sure that some linters are even able to identify accidental use outside tests.

Reply:
@ssbarnea I think at this point this PR needs a much bigger overhaul, and most of it may need to live in this repo. Probably best to hold off on cleaning up this specific code until the overhaul is done.